### PR TITLE
KOGITO-6067 : Tasks in BPMN and specifically BR Node do not allow type to be Collection with a generic

### PR DESCRIPTION
**Pull request by inodeman**
Hi @romartin can you check this PR? It included a lot of changes in several places due to the fact that we use commas to separate variables, I have tested it extensively and works fine. The only cases where it can fail is when a user enters malformed genercis, example Map<String,String, one thing that I can also add is if I detect that it is malformed, then not add it to the Data Type service to be used in other places, plus maybe adding some code to handle cases when it is malformed. Let me know

### CODE REVIEW

**Comment by inodeman** [comment]
> @inodeman my 2c, let's not allow our users to input malformed data types into the service, instead let's color the `input` red and notify him the data type is malformed.

Hi @domhanak yes, currently if the user enters it wrongly a pop up will appear, it is the same pop up that validates process/global variables, etc

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java
Comment (domhanak) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: What is this method supposed to do? It return null everytime.

Comment (inodeman) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: Hi @domhanak the code works so that null is expected to be a valid value, if not returns the validationError description, this method returns null as it is the base class, actual implementation is in sub class.

Comment (domhanak) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: If I understand this correctly, a string isBalanced when the returned value is null? Imho that is bad design and will lead to NPE's in the future. I suggest changing it so that the method actually returns `true` or `false` if the string is balanced. Here returning true.

Comment (inodeman) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: Hi @domhanak I understand your point, but I just made it work just like isValidValue method from subclasses. Look at this implementation from CustomDataTypeBox.java, let me know what you think
image

Comment (domhanak) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: Ok, I ll wait for @Josephblt review

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java
Comment (Josephblt) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java:31]: If you remove the comment below, please don't forget this import.

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java
Comment (Josephblt) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java:179]: I think you forgot to remove this comment here.

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java
Comment (Josephblt) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: I would make this method abstract and implement it on both CustomDataTypeTextBox and VariableNameTextBox.The same way isValidValue method does. 

I wouldn't worry about the true or false values. But I have to agree with @domhanak. The return types are confusing. Perhaps better naming of the methods to highlight their intent would improve things.

**Review by Josephblt — CHANGES_REQUESTED** [review]
Other than the command line and the abstract method, it is looking good to me. After those fixes, I can approve this.

**Comment by domhanak** [comment]
*Use cases*
I checked the vscode-extension and I was able to fulfill all required use cases. 
I was able to add generics as data types in process variables, global variables, imports and data assignments(both input/output), inputs and output of multiple instance sub-process.
Some examples of tested types:
`java.util.List<String>`
`java.util.Map<K,V>`
`java.util.Map<AgeDecisionService, AgeDecision>`
`java.util.Map<String,List<AgeDecision>>`

*Error handling*
For error handling, the error dialog was appearing in valid cases, but I kept losing focus of the custom data type input field.
Could we please improve it to not lose focus and keep the input field focused? That way user can correct the amount of `>``<` needed without clicking the drop-down, and then clicking Edit and then correcting the value.
This way it will also not be possible to input a malformed data type into the data type cache. 
What do you think? If we want to go one step further we can add the missing `<` `>` if possible. :)

*Multiple instance sub-process*
I noticed some strange behavior when adding a custom data types to the `MI Data Input` and `MI Data Output` data type input nodes. I was able to add a custom data types with generics, but this one issue kept appearing. It is a bit bad UX, it can also happen randomly when you are editting or adding new custom data type.

*Issue  desc. - Properties panel is reset when changing the MI data input/output datatypes*
Steps:
1. Create a process with multiple instance subprocess.
2. Configure some data types.
3. Select he MISubprocess node.
4. Open the MI input.output drop-down and click some data type and wait.
5. The properties panel reloads to main process.

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java
Comment (inodeman) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/widgets/AbstractValidatingTextBox.java:107]: Hi @Josephblt making it abstract and including GTLT in name to indicate balanced of Greater Than and Less Than

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java
Comment (inodeman) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java:179]: Thanks @Josephblt sorry about that

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java
Comment (inodeman) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/variablesEditor/VariablesEditorFieldRenderer.java:31]: Hi @Josephblt removed.

**Comment by inodeman** [comment]
Hi @domhanak thanks for review. I have modified it so that when displaying FQDN if it detects < or >, then it wont display the format with [] to prevent further issues. For the error handling issue you mention, if I understand correctly the issue is when you input and go to another field the pop up appears but the focus on the custom data type is lost? if so, it will be a big difficult to reset the focus to the previous element since the user clicked on another field, let me know if I am correct in my assumptions. The Multiple Instance issue you mention I tried to reproduce it during several minutes but did not see the issue, if possible can you share video? I am pushing the changes for the main issue, let me know if I am on the right track

**Review by Josephblt — APPROVED** [review]

**Comment by inodeman** [comment]
Hi @domhanak @romartin I made some changes, now input like ....... is not allowed, also empty generics <> or malformed generics >< is not allowed in input and as a result they do not get populated in listbox. Let me know

**Review by Josephblt — APPROVED** [review]

**Comment by yesamer** [comment]
jenkins retest this please

**Comment by romartin** [comment]
Hey @domhanak ,

Please whenever is possible, try to go ahead with this review. Thanks!

**Comment by inodeman** [comment]
jenkins retest this please

**Comment by domhanak** [comment]
I checked the vscode-extension and I was able to fulfill all required use cases.
I was able to add generics as data types in process variables, global variables, imports and data assignments(both input/output), inputs and output of multiple instance sub-process.

Some examples of tested types:
java.util.List<String>
java.util.Map<K,V>
java.util.Map<AgeDecisionService, AgeDecision>
java.util.Map<String,List<AgeDecision>>
java.util.Map<org.kie.businessapp.Age,String>
java.util.Map<java.util.List<String>,String>


Error handling is in a better state now, however it could be improved.
Reported minor: https://issues.redhat.com/browse/KOGITO-6427

**Review by domhanak — APPROVED** [review]

**Comment by domhanak** [comment]
Sorry, accidentally approved, while I just wanted to add another comment :)

I noticed one more issue:
1. Use a process with MultipleInstanceSubprocess
2. Add couple of process variables.
3. Add this process variable - name: "a", dataType: "java.util.Map<Age, List<Age>>" AND name: "a", dataType: "java.util.Map<Age, List<String>>"
4.  Navigate to multiple instance subprocess properties
5. Try to setup MI Collection output
6. You will see following options:
Selection_069
Screenshot from 2021-12-15 10-29-47

**Review by domhanak — CHANGES_REQUESTED** [review]
See latest comment, bug found.

**Comment by yesamer** [comment]
jenkins retest this please

**Comment by inodeman** [comment]
Hi @domhanak checking the issue, will update soon.

**Comment by inodeman** [comment]
Hi @domhanak fixed the issue, now you should only be seeing the process variable names, the other part, the data types should not be displayed, it was an issue with the commas that separate process variables. Will push changes shortly.

**Comment by domhanak** [comment]
Default build was cancelled, is there an issue with CI?

**Comment by domhanak** [comment]
jenkins retest this please

**Comment by inodeman** [comment]
Changes taken to new PR. https://github.com/kiegroup/kie-tools/pull/808

**Pull request closed without merging**
